File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:    
    int solverRec(TreeNode* root,int &maxSum){

        if(root==NULL){
            return 0;
        }

        int leftSum = max(0,solverRec(root->left,maxSum));
        int rightSum = max(0,solverRec(root->right, maxSum));

        maxSum = max(maxSum,leftSum+rightSum+root->val);
        return max(leftSum,rightSum)+root->val;
    }
public:
    int maxPathSum(TreeNode* root) {

        int maxSum =INT_MIN;
        solverRec(root,maxSum);
        return maxSum;
        
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    maxSum =-1e9
    def solverRec(self,root):
        if root ==None:
            return 0
        
        lSum = max(0,self.solverRec(root.left))
        
        rSum = max(0,self.solverRec(root.right))
        self.maxSum = max(self.maxSum, lSum+rSum+root.val)
        return max(lSum,rSum)+root.val
    def maxPathSum(self, root: Optional[TreeNode]) -> int:

        self.solverRec(root)
        return self.maxSum

        